sklearn/ensemble/gradient_boosting.py

```diff
@@ -1846,3 +1853,466 @@ def apply(self, X):
+            'min_weight_fraction_leaf': self.min_weight_fraction_leaf,
+            'max_depth': self.max_depth,
+            'init': self.init,
+            'max_features': self.max_features,
+            'max_leaf_nodes': self.max_leaf_nodes,
+            'presort': self.presort,
+        }
+
+        # Pre processing to ensure every parameter is a list
+        # which `ParameterGrid` can iterate over
+        for key, value in params.items():
+            params[key] = np.atleast_1d(value)
+
+        return params
+
+
+def _fit_single_param(estimator, X, y, train, test, params, stop_rounds,
+                      max_iter, scoring, score_precision):
+    """ Fit a single estimator till stopping criteria is reached.
+
+    Parameters
+    ---------
+    estimator : GradientBoostingRegressor or GradientBoostingClassifier
+        The class of the model to instantiate.
+
+    X : array-like, shape = [n_samples, n_features]
+        Training vectors, where n_samples is the number of samples
+        and n_features is the number of features.
+
+    y : array-like, shape = [n_samples]
+        Target values (integers in classification, real numbers in
+        regression)
+        For classification, labels must correspond to classes.
+
+    train : array-like
+        The indices of the training samples.
+
+    test : array-like
+        The indices of the testing samples.
+
+    params : dict
+        The parameters to instantiate the Gradient Boosting model with
+
+    stop_rounds : int
+        The number of iterations for which the score should remain constant
+        over to stop training.
+
+    max_iter : int
+        The maximum number of estimators to fit during gradient boosting.
+
+    scoring : string, callable or None, default=None
+        A string (see model evaluation documentation) or
+        a scorer callable object / function with signature
+        ``scorer(estimator, X, y)``.
+        If ``None``, the ``score`` method of the estimator is used.
+
+    score_precision : int
+        The number of decimal places the score during each iterations is
+        rounded off to before comparing with the previous ones.
+
+    Returns
+    -------
+    score: float
+        The score when the model converged.
+
+    n_iterations:
+        The number of iterations it took to converge. This is the same as
+        parameter `n_estimators` for the gradient boosting model
+    """
+
+    params['warm_start'] = True
+    gb = estimator(**params)
+    scorer = check_scoring(estimator, scoring=scoring)
+    scores = np.ones((stop_rounds,))
+
+    X_train = X[train]
+    y_train = y[train]
+    X_test = X[test]
+    y_test = y[test]
+
+    for i in range(1, max_iter):
+        gb.n_estimators = i
+        gb.fit(X_train, y_train)
+
+        scores = np.roll(scores, shift=-1)
+        scores[-1] = scorer(gb, X_test, y_test)
+        rscores = np.round(scores, score_precision)
+
+        if np.allclose(rscores, rscores[-1]):
+            break
```

Comment:
After making this change - https://github.com/scikit-learn/scikit-learn/pull/5689/files#r72066046